### PR TITLE
Raw to digits

### PR DESCRIPTION
**Pull request by cortesep**

### CODE REVIEW

**Comment by shahor02** [comment]
there were also plenty of `Double_t`...

**Comment by shahor02** [comment]
@cortesep wait, I see you have same commits in this PR and in the earlier https://github.com/AliceO2Group/AliceO2/pull/5533/files. By amending both you will introduce conflicts.

I would suggest to do
````
git checkout dev_zdcawrw
git rebase working_branch
````
Add all fixes, make sure that it compiles and works both for mc->raw and raw->digis, the forcepush by 
`git push -f cortesep:working_branch`

Then the old one can be closed.

**Comment by cortesep** [comment]
Hi Ruben,

On 3/1/21 11:55 AM, Ruben Shahoyan wrote:
> @cortesep <https://github.com/cortesep> wait, I see you have same 
> commits in this PR and in the earlier 
> https://github.com/AliceO2Group/AliceO2/pull/5533/files 
> <https://github.com/AliceO2Group/AliceO2/pull/5533/files>. By amending 
> both you will introduce conflicts.
> 
> I would suggest to do
> 
> |git checkout dev_zdcawrw git rebase working_branch |

I'm at this stage. What should I do now? Everything seemed to run smoothly..


git status
# On branch working_branch
nothing to commit, working directory clean
[cortese@fis01 O2]$ git checkout dev_zdcawrw
Switched to branch 'dev_zdcawrw'
[cortese@fis01 O2]$ git rebase working_branch
First, rewinding head to replay your work on top of it...
Fast-forwarded dev_zdcawrw to working_branch.

Can I do the following step?

**Comment by shahor02** [comment]
@cortesep yes, if all the fixes (UShort_t etc) and comments are accounted in this one, just do ``push -f ...``

**Comment by cortesep** [comment]
Hi,

I fixed also the Short..

Now I get this error message:

git push -f cortesep:working_branch
fatal: You are pushing to remote 'cortesep:working_branch', which is not 
the upstream of
your current branch 'dev_zdcawrw', without telling me what to push
to update which remote branch.

What do you suggest?
Thanks,
Pietro

**Comment by shahor02** [comment]
You don't need `:` in the push, just ``git push -f cortesep working_branch``

File: Detectors/ZDC/simulation/src/Digits2Raw.cxx [hunk 1]
```diff
@@ -412,16 +325,16 @@ void Digits2Raw::insertLastBunch(int ibc, uint32_t orbit)
...
       if (mModuleConfig->modules[im].readChannel[ic]) {
         auto id = mModuleConfig->modules[im].channelID[ic];
         auto base_m = mSimCondition->channels[id].pedestal;      // Average pedestal
         auto base_s = mSimCondition->channels[id].pedestalFluct; // Baseline oscillations
         auto base_n = mSimCondition->channels[id].pedestalNoise; // Electronic noise
-        Double_t base = gRandom->Gaus(base_m, base_s);
-        Int_t is = 0;
-        Double_t val = base + gRandom->Gaus(0, base_n);
+        double base = gRandom->Gaus(base_m, base_s);
+        int32_t is = 0;
```
Comment (shahor02) [inline Detectors/ZDC/simulation/src/Digits2Raw.cxx]: again unused `is` which is then incremented

File: Detectors/ZDC/simulation/src/Digits2Raw.cxx [hunk 2]
```diff
@@ -494,15 +406,15 @@ void Digits2Raw::convertDigits(int ibc)
...
     // This is a limitation of software but we are not supposed to acquire the
     // same signal twice anyway
-    for (Int_t im = 0; im < NModules; im++) {
-      for (UInt_t ic = 0; ic < NChPerModule; ic++) {
+    for (int32_t im = 0; im < NModules; im++) {
+      for (uint32_t ic = 0; ic < NChPerModule; ic++) {
         if (mModuleConfig->modules[im].channelID[ic] == chd.id &&
             mModuleConfig->modules[im].readChannel[ic]) {
-          Int_t is = 0;
+          int32_t is = 0;
```
Comment (shahor02) [inline Detectors/ZDC/simulation/src/Digits2Raw.cxx:405]: this `is` is used, but usually one does e.g. `mZDC.data[im][ic].f.s00 = chd.data[is++];`

**Review by shahor02 — COMMENTED** [review]
OK, thanks, I am testing the chain. See some comments below.
There are still some root types left, see the result of ``~/alice/O2/Detectors/ZDC$ git grep '_t' | grep -E 'Int|Short|Float|Double'``

File: Detectors/ZDC/simulation/src/Digits2Raw.cxx (hunk 1 above)
Comment (cortesep) [reply Detectors/ZDC/simulation/src/Digits2Raw.cxx]: It's a leftover from previous code. Thanks for spotting this.

**Comment by shahor02** [comment]
Hi @cortesep 

While checking with 100 pbpb events, I see in the log:
````
[1247062:zdc-datareader-dpl]: [15:03:50][INFO] ZDCDataReaderDPLSpec
[1247062:zdc-datareader-dpl]: [15:03:50][INFO] Pages: 8206
[1247062:zdc-datareader-dpl]: [15:03:50][INFO] Processing #bc 232
[1247062:zdc-datareader-dpl]: [15:03:50][INFO] Number of Digits: 232
[1247062:zdc-datareader-dpl]: [15:03:50][INFO] Number of ChannelData: 4600
[1247062:zdc-datareader-dpl]: [15:03:50][INFO] Number of PedestalData: 26
````
while in the simulated digits I see less entries:
````
TFile *_file0 = TFile::Open("zdcdigits.root")
o2sim->Scan("@ZDCDigitCh.size() : @ZDCDigitBC.size()")
************************************
*    Row   * @ZDCDigit *  @ZDCDigi *
************************************
*        0 *      3924 *       206 *
************************************
````
So, there is a difference between the input and decoded data.
But the number of decoded digits = N of simulated digits + N of pedestals and the
difference in the number of decoded and simulated channels / Npedestals is (4600-3924)/26 = 26, i.e. the number of channels.
I guess it is not normal, right? 
Also, somehow I don't see decoded digits written, checking why.

A general question: what is the reason to split the decoder to ``RawReaderBase -> RawReaderZDCBase -> RawReaderZDC``, and on top of that to make ``ZDCDataReaderDPLSpec`` take the RawReader templated parameter? Can't this be just 1 decoder being a data member of ZDCDataReaderDPLSpec?

**Comment by shahor02** [comment]
PS ignore  ``I don't see decoded digits written``, I see that this due to the typo: you are sending the channels twice but don't send pedestals.

**Review by shahor02 — CHANGES_REQUESTED** [review]
1 )Comparing the content of decoded and original digits, I see that ZDCDigitBC in decoded has many fields not set, e.g. references on channels, or different from original ones (triggers):

original digits:
````
root [46] d0->Scan("ZDCDigitBC.print()")
***********************************
*    Row   * Instance * ZDCDigitB *
***********************************
Orbit         1 bc  233 nch 26 pos 0
Read: 0[R RR] 1[ RRR] 2[R RR] 3[ RRR] 4[RRRR] 5[ RRR] 6[RRRR] 7[ RRR]
Hits: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
AUTO: 0 --1-- 1 --1-- 2 --1-- 3 --1-- 4 --1-- 5 --1-- 6 --1-- 7 --1--
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
*        0 *        0 *         0 *
Orbit         1 bc  234 nch 26 pos 26
Read: 0[R RR] 1[ RRR] 2[R RR] 3[ RRR] 4[RRRR] 5[ RRR] 6[RRRR] 7[ RRR]
Hits: 0[HHHH] 1[HHHH] 2[HHHH] 3[HHHH] 4[HHHH] 5[HH  ] 6[HHHH] 7[HHH ]
AUTO: 0 ---0- 1 ---0- 2 ---0- 3 ---0- 4 ---0- 5 ---0- 6 ---0- 7 ---0-
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
*        0 *        1 *         0 *
Orbit         1 bc  363 nch  3 pos 52
Read: 0[    ] 1[    ] 2[    ] 3[ RRR] 4[    ] 5[    ] 6[    ] 7[    ]
Hits: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
AUTO: 0 ----- 1 ----- 2 ----- 3 --1-- 4 ----- 5 ----- 6 ----- 7 -----
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
*        0 *        2 *         0 *
Orbit         1 bc  364 nch  3 pos 55
Read: 0[    ] 1[    ] 2[    ] 3[ RRR] 4[    ] 5[    ] 6[    ] 7[    ]
Hits: 0[    ] 1[    ] 2[    ] 3[HH  ] 4[    ] 5[    ] 6[    ] 7[    ]
AUTO: 0 ----- 1 ----- 2 ----- 3 ---0- 4 ----- 5 ----- 6 ----- 7 -----
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
...
````
decoded digits:
````root [47] d1->Scan("ZDCDigitBC.print()")
***********************************
*    Row   * Instance * ZDCDigitB *
***********************************
Orbit         1 bc  233 nch  0 pos 0
Read: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
Hits: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
AUTO: 0 --1-- 1 --1-- 2 --1-- 3 --1-- 4 --1-- 5 --1-- 6 --1-- 7 --1--
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
*        0 *        0 *         0 *
Orbit         1 bc  234 nch  0 pos 0
Read: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
Hits: 0[ HHH] 1[H HH] 2[ HHH] 3[HHHH] 4[ H  ] 5[HHHH] 6[ HH ] 7[H HH]
AUTO: 0 ---0- 1 ---0- 2 ---0- 3 ---0- 4 ---0- 5 ---0- 6 ---0- 7 ---0-
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
*        0 *        1 *         0 *
Orbit         1 bc  363 nch  0 pos 0
Read: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
Hits: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
AUTO: 0 ----- 1 ----- 2 ----- 3 --1-- 4 ----- 5 ----- 6 ----- 7 -----
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
*        0 *        2 *         0 *
Orbit         1 bc  364 nch  0 pos 0
Read: 0[    ] 1[    ] 2[    ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
Hits: 0[    ] 1[    ] 2[ H  ] 3[    ] 4[    ] 5[    ] 6[    ] 7[    ]
AUTO: 0 ----- 1 ----- 2 ----- 3 ---0- 4 ----- 5 ----- 6 ----- 7 -----
ALIT: 0 ----  1 ----  2 ----  3 ----  4 ----  5 ----  6 ----  7 ---- 
````

2) could you rename ``o2-zdc-flp-dpl-workflow`` to "o2-zdc-raw2digits``, I am not sure this can run on flp.

3) since you are using ccdb in the decoding, please add an option to set the url.

4) there is still a conflict in options of mc->raw:
````
o2-zdc-digi2raw -h
...
  -o [ --output-dir ] arg (=./)         output directory for raw data
  -o [ --ccdb-url ] arg                 url of the ccdb repository
````
e.g. change in ``digi2raw.cxx``
````
add_option("ccdb-url,o", bpo::value<std::string>()->default_value(""), "url of the ccdb repository")
````
to
````
add_option("ccdb-url,c", bpo::value<std::string>()->default_value(""), "url of the ccdb repository")
````

**Comment by cortesep** [comment]
Hi,
could you send me the modifications to write digits on file?
I will have a look to the discrepancies.
Pietro

**Comment by shahor02** [comment]
More problems:
I've digitized the same 100 events with 5kHz and converted to raw with 64-orbit TFs, leading to 4 TFs created. But in decoding I see just 1 output from the ``accumulateDigits``

````
o2-sim-digitizer-workflow --interactionRate 5000
o2-zdc-digi2raw  --output-dir raw/ZDC --configKeyValues "HBFUtils.nHBFPerTF=64"
o2-raw-file-reader-workflow --configKeyValues "HBFUtils.nHBFPerTF=64"  --input-conf raw/ZDC/ZDCraw.cfg  | o2-zdc-flp-dpl-workflow
...
[1290335:raw-file-reader]: [16:20:28][INFO] Sent payload of 269792 bytes in 16 parts in 1 messages for TF 0 | Timing (total/IO): 0.000e+00 / 0.000e+00
[1290335:raw-file-reader]: [16:20:28][INFO] Reading TF#1 (1 at iteration 0)
[1290335:raw-file-reader]: [16:20:28][INFO] Sending 16 parts to channel from_raw-file-reader_to_zdc-datareader-dpl
[1290335:raw-file-reader]: [16:20:28][INFO] Sent payload of 271904 bytes in 16 parts in 1 messages for TF 1 | Timing (total/IO): 0.000e+00 / 0.000e+00
[1290335:raw-file-reader]: [16:20:28][INFO] Reading TF#2 (2 at iteration 0)
[1290335:raw-file-reader]: [16:20:28][INFO] Sending 16 parts to channel from_raw-file-reader_to_zdc-datareader-dpl
[1290335:raw-file-reader]: [16:20:28][INFO] Sent payload of 264192 bytes in 16 parts in 1 messages for TF 2 | Timing (total/IO): 0.000e+00 / 0.000e+00
[1290335:raw-file-reader]: [16:20:28][INFO] Reading TF#3 (3 at iteration 0)
[1290335:raw-file-reader]: [16:20:28][INFO] Sending 16 parts to channel from_raw-file-reader_to_zdc-datareader-dpl
[1290335:raw-file-reader]: [16:20:28][INFO] Sent payload of 168032 bytes in 16 parts in 1 messages for TF 3 | Timing (total/IO): 0.000e+00 / 0.000e+00
[1290335:raw-file-reader]: [16:20:28][INFO] Finished: payload of 973920 bytes in 64 messages sent for 4 TFs
[1290335:raw-file-reader]: [16:20:28][INFO] Timing for            Init: Cpu: 0.000e+00 Real: 6.618e-04 s in 0 slots
...
````
I put data for both tests here: https://cernbox.cern.ch/index.php/s/3Is4aPb4ZGC2eLM
````
tar cvzf zdcdata.tar.gz dig50khz dig5khz
dig50khz/
dig50khz/o2sim_geometry.root
dig50khz/o2sim_grp.root
dig50khz/zdcdigits.root
dig50khz/raw/
dig50khz/raw/ZDC/
dig50khz/raw/ZDC/ZDCraw.cfg
dig50khz/raw/ZDC/zdc.raw
dig5khz/
dig5khz/o2sim_geometry.root
dig5khz/o2sim_grp.root
dig5khz/zdcdigits.root
dig5khz/raw/
dig5khz/raw/ZDC/
dig5khz/raw/ZDC/ZDCraw.cfg
dig5khz/raw/ZDC/zdc.raw
````

**Comment by shahor02** [comment]
````
--- a/Detectors/ZDC/workflow/include/ZDCWorkflow/RawReaderZDC.h
+++ b/Detectors/ZDC/workflow/include/ZDCWorkflow/RawReaderZDC.h
@@ -70,7 +70,7 @@ class RawReaderZDC : public RawReaderZDCBase
   {
     pc.outputs().snapshot(o2::framework::Output{o2::header::gDataOriginZDC, "DIGITSBC", 0, o2::framework::Lifetime::Timeframe}, mDigitsBC);
     pc.outputs().snapshot(o2::framework::Output{o2::header::gDataOriginZDC, "DIGITSCH", 0, o2::framework::Lifetime::Timeframe}, mDigitsCh);
-    pc.outputs().snapshot(o2::framework::Output{o2::header::gDataOriginZDC, "DIGITSCH", 0, o2::framework::Lifetime::Timeframe}, mPedestalData);
+    pc.outputs().snapshot(o2::framework::Output{o2::header::gDataOriginZDC, "DIGITSPD", 0, o2::framework::Lifetime::Timeframe}, mPedestalData);
   }
   bool mDumpData;
 };
````

**Comment by cortesep** [comment]
Made some checks and raw converted into digits correspond to original inputs with some notes
1) not all raw data is converted back to digits but only a TF
2) the new digits have the addition of pedestal sample data that are not present in simulated digits
3) the simulated digits contain some elements of MC truth
  - the information of hits in channels that are not readout
  - the information of triggers in channels that are not readout
Consequently the hit/trigger bit pattern may differ from original pattern for channels that are not readout

**Comment by cortesep** [comment]
A more general question. The scaler data is not propagated into digits, therefore using digits we don't have the possibility to estimate the luminosity from this source (useful in triggered mode, redundant in autotrigger mode).
One could add also this information, however it will depend also on which information will be archived in RUN3 for ZDC

**Comment by shahor02** [comment]
Hi @cortesep,


>    not all raw data is converted back to digits but only a TF

This is certainly a problem, will study.

>    the new digits have the addition of pedestal sample data that are not present in simulated digits

you did this on purpose, so this is as expected, right?

>    the simulated digits contain some elements of MC truth
>    the information of hits in channels that are not readout
>    the information of triggers in channels that are not readout
>    Consequently the hit/trigger bit pattern may differ from original pattern for channels that are not readout

I did not understand if all subitems are ok or not. Why the digit should contain MC truth at all? 
What is it used for (the MC label is different story). Should not we reduce the info in simulated digit so that it provides the same info as decoded one?

> A more general question. The scaler data is not propagated into digits, therefore using digits we don't have the possibility to estimate the luminosity from this source (useful in triggered mode, redundant in autotrigger mode).
> One could add also this information, however it will depend also on which information will be archived in RUN3 for ZDC

For the scalers: was not it decided that the ZDC will be only in auto-triggered mode? What do you mean by what is archived? For the raw data, as we agreed, I've already implemented the CTF for ZDC digits.
What would be the extra scalers size? Unless it is really needed, I would prefer to not inflate the data size.

**Comment by cortesep** [comment]
Hi,

> I did not understand if all subitems are ok or not. Why the digit should 
> contain MC truth at all?
> What is it used for (the MC label is different story). Should not we 
> reduce the info in simulated digit so that it provides the same info as 
> decoded one?

During digitization I was flagging the channel that caused the module to 
be readout (for the purpose of digitizing it). However in raw data you 
don't have this information for a trigger channel that is not readout.

This info is nevertheless useful for debugging. I should probably by 
default suppress this information and introduce a configuration 
parameter to allow to enable it.

> 
>     A more general question. The scaler data is not propagated into
>     digits, therefore using digits we don't have the possibility to
>     estimate the luminosity from this source (useful in triggered mode,
>     redundant in autotrigger mode).
>     One could add also this information, however it will depend also on
>     which information will be archived in RUN3 for ZDC
> 
> For the scalers: was not it decided that the ZDC will be only in 
> auto-triggered mode? What do you mean by what is archived? For the raw 
> data, as we agreed, I've already implemented the CTF for ZDC digits.
> What would be the extra scalers size? Unless it is really needed, I 
> would prefer to not inflate the data size.

The ZDC will run in autotrigger mode but in principle can run also in 
triggered mode, and in this situation we will lack the information about 
ZDC event rate, unless we add it to the digits.

Sorry that I didn't realize before it was missing. The information is 
available for each channel at each orbit (uint16_t) so the space is not 
big and it should be easily compressible because all the orbits will 
have similar counts. Probably one could eventually integrate it on a TF.
What do you think?

Ciao,
Pietro

**Comment by shahor02** [comment]
> Sorry that I didn't realize before it was missing. The information is
available for each channel at each orbit (uint16_t) so the space is not
big and it should be easily compressible because all the orbits will
have similar counts. Probably one could eventually integrate it on a TF.

If it is the same counter for all channels, why don't you add just 1 scaler to BCData? 
For the one integrated over TF: currently we don't have any ZDC object created once per TF, I would not introduce it just for the sake of a single scaler.

**Comment by cortesep** [comment]
Hi,

it is a good place, however it would be

std::array<uint16_t, NChannels> scaler;

The information would be needed to fill this:
O2/DataFormats/Detectors/ZDC/include/DataFormatsZDC/OrbitRecData.h
in case we are not in autotrigger mode.

Pietro

**Comment by shahor02** [comment]
> it is a good place, however it would be std::array<uint16_t, NChannels> scaler;

Wait, if you need it per orbit, there is no point in adding it per digit. Why don't you add it to PedestalData ?

**Comment by cortesep** [comment]
Done.. at least for raw->digits conversion that is what we need.
Probably PedestalData should be renamed but this involves many files.
Should I go on with the modifications?

Ciao,
Pietro

**Comment by shahor02** [comment]
@cortesep , 
to rename the ``PedestalData`` to e.g. ``PedestalScalerData`` (or ``EndOfOrbitData``, you decide), just do from O2 dir.:
````
git grep -l PedestalData | grep -i zdc | xargs perl -pi -e 's/PedestalData/PedestalScalerData/g'
````

Did you validate the changes on 2 PbPb data samples I put on cernbox? Is the content and the number of original and decoded digits now consistent?

**Comment by cortesep** [comment]
Hi,

I was not able to run the digitization on PbPb data. For some reasons 
the digits are empty. I was able to run it on emd simulation.

The entries are now coherent.
There is a new switch to store also the samples with baseline of bunch 3564

o2-zdc-raw2digits --not-check-trigger

in Digitizer class there is now

   bool mMaskTriggerBits = true;                                  // 
Mask trigger bits with readout mask

To hide the hit information that is known only to simulation and is not 
available in raw data. I didn't know how to set this through 
ZDCDigitizerSpec

ZDCDigitizerSpec is calling a new function "Finalize" to cleanup the 
information and sum the scalers. However in simulated data this is 
limited to triggering channels, while if you convert from raw you get 
also the scalers of non triggering channels.

After renaming PedestalData to OrbitData it should be finished..

Pietro

**Comment by shahor02** [comment]
> To hide the hit information that is known only to simulation and is not available in raw data. I didn't know how to set this through
ZDCDigitizerSpec, get the flag in its ``init()`` method, like https://github.com/AliceO2Group/AliceO2/blob/0ad5561d6a1663593b4d81a3772ded13e9f7ab4c/Steer/DigitizerWorkflow/src/ITSMFTDigitizerSpec.cxx#L69 and then set it to Digitizer. I would suggest to not store this info by defaults, i.e. the option should be something like
````
Options{{"enable-hit-info", o2::framework::VariantType::Bool, false, {"enable hit info of unread channels"}}
````

BTW, I would suggest also to propagate the ``useMC`` flag (``--disable-mc`` of ``o2-sim-digitizer-workflow``) which will allow to skip creation of labels info, see: https://github.com/AliceO2Group/AliceO2/blob/0ad5561d6a1663593b4d81a3772ded13e9f7ab4c/Steer/DigitizerWorkflow/src/ITSMFTDigitizerSpec.cxx#L334

Once you finish, please remove [WIP]. You could not redigitize pbpb because I did not copy Hits and Kine, they are too boog. I'll redigitize both 1 and 4 TF versions and send to you for final validation.


https://github.com/AliceO2Group/AliceO2/blob/0ad5561d6a1663593b4d81a3772ded13e9f7ab4c/Steer/DigitizerWorkflow/src/ITSMFTDigitizerSpec.cxx#L332

**Comment by cortesep** [comment]
> > To hide the hit information that is known only to simulation and is not available in raw data. I didn't know how to set this through
> > ZDCDigitizerSpec, get the flag in its `init()` method, like https://github.com/AliceO2Group/AliceO2/blob/0ad5561d6a1663593b4d81a3772ded13e9f7ab4c/Steer/DigitizerWorkflow/src/ITSMFTDigitizerSpec.cxx#L69
> >  and then set it to Digitizer. I would suggest to not store this info by defaults, i.e. the option should be something like
> 
> ```
> Options{{"enable-hit-info", o2::framework::VariantType::Bool, false, {"enable hit info of unread channels"}}

Done.

> 
> BTW, I would suggest also to propagate the `useMC` flag (`--disable-mc` of `o2-sim-digitizer-workflow`) which will allow to skip creation of labels info, see:

I have an issue with disable-mc options. If I introduce it in the list of options
  return DataProcessorSpec{
    "ZDCDigitizer",
    Inputs{InputSpec{"collisioncontext", "SIM", "COLLISIONCONTEXT", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe}},
    outputs,
    AlgorithmSpec{adaptFromTask<ZDCDPLDigitizerTask>()},
    Options{{"enable-hit-info", o2::framework::VariantType::Bool, false, {"enable hit info of unread channels"}},
            {"disable-mc", o2::framework::VariantType::Bool, false, {"do not store MC labels"}}}
    };
I have an error
Error: option '--disable-mc' is ambiguous and matches different versions of '--disable-mc'

If the option is not in the list the program crashes..

**Comment by cortesep** [comment]
I used a trick to set an option coherent with mctruth argument. I don't know if this is what you meant but it seems to work

**Comment by davidrohr** [comment]
I think we should not use an option with a different name here (`disable-mc-zdc`), but it should use the same `disable-mc` option of all other workflows, otherwise it is misleading and failure-prone.

I checked what the TPC workflow is doing, and it defines the `disable-mc` option:
```
void customize(std::vector<o2::framework::ConfigParamSpec>& workflowOptions)
{
  using namespace o2::framework;
  std::vector<ConfigParamSpec> options{
[...]
    {"disable-mc", VariantType::Bool, false, {"disable sending of MC information"}},
[...]
  };
  std::swap(workflowOptions, options);
}
```

I am wondering why it should be ambiguous here. Due to the different option description perhaps?
```
{"disable-mc-zdc", o2::framework::VariantType::Bool, false, {"do not store ZDC MC labels"}}}
```
But in any case, I think we should understand and fix that.

**Comment by cortesep** [comment]
Only remaining difference is that Digitizer writes a single entry for all events, while raw to digits writes one entry per TF

**Comment by shahor02** [comment]
@cortesep You don't need to add `disable-mc` on your digitizer device level, it is already defined on the workflow level: https://github.com/AliceO2Group/AliceO2/blob/e61b69ee20de242e9bcf0da86ee3c3a776d7599f/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx#L163 , https://github.com/AliceO2Group/AliceO2/blob/e61b69ee20de242e9bcf0da86ee3c3a776d7599f/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx#L383-L384 and is passed to every device, e.g. https://github.com/AliceO2Group/AliceO2/blob/e61b69ee20de242e9bcf0da86ee3c3a776d7599f/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx#L460-L462

So, you just need to add to getZDCDigitizerSpec and getZDCDigitWriterSpec an argument ``(bool usMC)`` and if it is on, skip everything related to MC labels, like others do.

**Comment by shahor02** [comment]
> Only remaining difference is that Digitizer writes a single entry for all events, while raw to digits writes one entry per TF
This is by design.

**Comment by cortesep** [comment]
Didn't realize that I could change the constructor. Now it should be ok.

File: Detectors/ZDC/simulation/src/Digitizer.cxx [hunk 3]
```diff
@@ -551,27 +606,70 @@ void Digitizer::assignTriggerBits(uint32_t ibc, std::vector<BCData>& bcData)
...
+    }
+  }
+  if (mMaskTriggerBits) {
+    for (uint32_t ibc = 0; ibc < bcData.size(); ibc++) {
+      auto& currBC = bcData[ibc];
+      for (int im = 0; im < NModules; im++) {
+        // Cleanup hits if module has no trigger
+        uint32_t mmask = 0xf << im;
+        if ((currBC.triggers & mmask) == 0) {
+          currBC.triggers&(~mmask);
```
Comment (shahor02) [inline Detectors/ZDC/simulation/src/Digitizer.cxx]: this has no effect and macOS CI fails on this. What should be here?

**Review by shahor02 — CHANGES_REQUESTED** [review]
@cortesep  please fix this 
````
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5585/0/DataFormats/Detectors/ZDC/src/BCData.cxx:19:17: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5585/0/DataFormats/Detectors/ZDC/src/BCData.cxx:21:7: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5585/0/DataFormats/Detectors/ZDC/src/BCData.cxx:39:12: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5585/0/DataFormats/Detectors/ZDC/src/BCData.cxx:67:12: error: statement should be inside braces [readability-braces-around-statements]
/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5585/0/DataFormats/Detectors/ZDC/src/BCData.cxx:75:12: error: statement should be inside braces [readability-braces-around-statements]
````
and see 1 comment below.

Also, the ``Steer/DigitizerWorkflow/src/ZDCDigitWriterSpec.h`` practically duplicates the ``Detectors/ZDC/workflow/src/ZDCDigitWriterDPLSpec.cxx``.
Could you add to the latter the optional MC output, eliminate the one in the ``Steer/DigitizerWorkflow``? Just call it here: https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Detectors/ZDC/workflow/src/o2-zdc-raw2digits.cxx#L59 as:

``specs.emplace_back(o2::zdc::getZDCDigitWriterDPLSpec(false));``

File: Detectors/ZDC/simulation/src/Digitizer.cxx (hunk 3 above)
Comment (cortesep) [reply Detectors/ZDC/simulation/src/Digitizer.cxx]: 1) Done
2) I see. However one is for digits from simulation and the other is for digits from raw. Should I merge the two into one file? Shouldn't the output file name be different for sim->digit and raw->digit?
3) It was a bug.. It's not a common condition, probably it didn't show up in the events that I tested.

Comment (shahor02) [reply Detectors/ZDC/simulation/src/Digitizer.cxx]: For the DigitWriter: yes, the names should be different, but they (actually, the default, which is "zdcdigits.root" for digitization and  "o2digit_zdc.root" for decoding) can be passed as an argument. So, the https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Detectors/ZDC/workflow/src/ZDCDigitWriterDPLSpec.cxx#L28-L41 should look like 

````
template <typename T>
using BranchDefinition = MakeRootTreeWriterSpec::BranchDefinition<T>;
DataProcessorSpec getZDCDigitWriterDPLSpec(bool mctruth, bool simVersion)
{
  using InputSpec = framework::InputSpec;
  using MakeRootTreeWriterSpec = framework::MakeRootTreeWriterSpec;
  std::string fnameDef = simVersion ? "zdcdigits.root" : o2digit_zdc.root";
  std::string writerName = simVersion ? "ZDCDigitWriterSim" : "ZDCDigitWriterSimDec";

  return MakeRootTreeWriterSpec(
 writerName,  fnameDef,  "o2sim", 1, 
 BranchDefinition<std::vector<o2::zdc::BCData>>{InputSpec{"digitBCinput", "ZDC", "DIGITSBC"}, "ZDCDigitBC"},
 BranchDefinition<std::vector<o2::zdc::ChannelData>>{InputSpec{"digitChinput", "ZDC", "DIGITSCH"}, "ZDCDigitCh"},
 BranchDefinition<std::vector<o2::zdc::OrbitData>>{InputSpec{"digitPDinput", "ZDC", "DIGITSPD"}, "ZDCDigitOrbit"},
 BranchDefinition<o2::dataformats::MCTruthContainer<o2::zdc::MCLabel>>{InputSpec{"labelinput", "ZDC", "DIGITSLBL"},  "ZDCDigitLabels", mctruth ? 1 : 0})(); 
}
````

Then, call it on https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Steer/DigitizerWorkflow/src/SimpleDigitizerWorkflow.cxx#L526 as 
````
specs.emplace_back(o2::zdc::getZDCDigitWriterDPLSpec(mctruth, true));
````
and  here https://github.com/AliceO2Group/AliceO2/blob/d99f406de1020f095ebfa69e7c30d67bf8f02179/Detectors/ZDC/workflow/src/o2-zdc-raw2digits.cxx#L59
as 
````
specs.emplace_back(o2::zdc::getZDCDigitWriterDPLSpec(false, false));
````

Comment (cortesep) [reply Detectors/ZDC/simulation/src/Digitizer.cxx]: Merged functionality into ZDCDigitWriterDPLSpec

**Review by shahor02 — APPROVED** [review]
@cortesep OK, I think we can squash/merge it already, if fixes will be needed, will do them in separate PR.

**Comment by shahor02** [comment]
No, there are conflicts, could you rebase to fresh dev and resolve the conflicts?

**Comment by davidrohr** [comment]
Hi @shahor02  there are only rebase conflicts, you can squash merge already now.
Rebasing doesn't work since apparently there were merge commits in between, that solved some conflicts in between, so just rebasing the individual commits is not possible.

**Comment by shahor02** [comment]
@davidrohr right, thanks!

**Pull request merged**